### ISSUE TITLE
Extract log

### ISSUE THREAD

**Issue by PierreDemailly** [MyUnisoft/sigyn#50]
The idea is to be able to extract info from the logs and reuse them in the alert template.

**First step**
Extract a label
Given this rule
```json
{
  "name": "[PROD] SyntaxError",
  "logql": "{env=\"production\"} |= `SyntaxError: Unexpected token`",
  "extractLabels": ["app"]
  "alert": {
    "on": {
      "count": 1,
      "interval": "5m"
    },
    "template": {
      "title": "{ruleName} ({label.app})",
      "content": [
        "- LogQL: {logql}"
      ]
    }
  }
}
```

Here, `label.app` is not supported, it has to be extracted from the fetched logs from `extractLabels` (juste an example, need to find a way to configure this)

> **Note** What to do if `count` > 1 ? It could have multiple app.

**Next**
Extract anything in a log, not only labels.

**Comment by fraxken** [comment]
@PierreDemailly maybe we should assign `unknown` if we cannot found value for label.app

**Comment by fraxken** [comment]
Probably need a JSON a bit more neutral (if we want to add more later like line extraction).

```json
"extract": {
  "labels": ["app"],
  "line": {
    "fieldName": "...RegExp here..."
  }
}
```

**Comment by PierreDemailly** [comment]
> Probably need a JSON a bit more neutral (if we want to add more later like line extraction).
> 
> ```json
> "extract": {
>   "labels": ["app"],
>   "line": {
>     "fieldName": "...RegExp here..."
>   }
> }
> ```

Yeah, so given your example, it would be `{line.fieldName}` to display the extracted fieldName somewhere ?

**Comment by fraxken** [comment]
Yes,

However I was thinking maybe we don't need to add extra configuration to tell to extract labels. If we decide that the default value if there is no label is unknown.

**Comment by PierreDemailly** [comment]
```json
{
  "name": "[PROD] SyntaxError",
  "logql": "{env=\"production\"} |= `SyntaxError: Unexpected token`",
  "alert": {
    "on": {
      "count": 10,
      "interval": "5m"
    },
    "template": {
      "title": "{ruleName} ({label.app})",
      "content": [
        "- LogQL: {logql}"
      ]
    }
  }
}
```
Given this JSON, note the `count` now set to `10`.

This LogQL is probably not accurate as a `SyntaxError` should be scoped to an app, but imagine we could find 10 logs within 10 differents app so how handle it ?

Using  `join(', ')` seems good in first time but we need to store labels in DB.

**Comment by PierreDemailly** [comment]
I'll introduce 2 news tables: `ruleLogs` (which will anyway be usefull later for #47) and `logLabels`

**Comment by fraxken** [comment]
@PierreDemailly Maybe not a definitive answer, but for me if there is 10 diffent labels app then `unknown` remain accurate.

I guess we should avoid going with something too complicated at first.

**Comment by PierreDemailly** [comment]
Label extract is done (#58)

**Comment by PierreDemailly** [comment]
As discussed, we can *easily* extract lines and create labels with `| regexp` [parser](https://grafana.com/docs/loki/latest/logql/log_queries/#regular-expression).

``{env=\"prod\", app=\"foo\"} |= `response returned \"POST /api/foo\"` | regexp `\\((?P<responseTime>\\d+\\.\\d+ms)\\)` ``

Now `{label.responseTime}` is available anywhere in the template (`title`, `content`)

We can still add config to simplify this, for instance adding a `regexp` field that takes an array of regexp to simplify the `logql` one.

**Comment by PierreDemailly** [comment]
@fraxken Should we close this issue ?
I've looked for a ES RegExp -> RE2 RegExp for a new `logql.regexp` field but it look too hard to implements and I'm not sure for the plus-value anyway (even worse if the field requires RE2 syntax)
I don't have other idea, do you ?

**Comment by fraxken** [comment]
No I think we are good to close now!
